Databricks SQL operators

alexott opened the pull request:
This PR adds new operators to Databricks provider:
* `DatabrickSqlOperator` that allows to execute SQL commands against Databricks SQL Endpoints and Databricks clusters.
* `DatabricksCopyIntoOperator` (built on top of `DatabrickSqlOperator`) that allows to import data into Databricks tables.

This operator uses the same connection as other Databricks operators (although it could be discussed), if having a dedicated connection make sense as we can further customize it with specific input fields, etc.

Another possible improvement - make the `databricks-sql-connector` dependency optional, but I'm not sure how to make it correctly in Airflow

closes: #21030 
closes: #21376

alexott committed:
Initial version of Databricks SQL operator

No documentation & tests yet

alexott committed:
Added documentation

Still need to fix existing tests & add tests for Databricks SQL hook &
operator

pateash commented on airflow/providers/databricks/hooks/databricks_base.py:
```diff
@@ -0,0 +1,289 @@
+DEFAULT_DATABRICKS_SCOPE = "2ff814a6-3304-4ab8-85cb-cd0e6f879c1d"
+
+
+class DatabricksBaseHook(BaseHook):
```
Is it possible to use this Hook directly, if not it should be named
BaseDatabricksHook, just to be consistent with other base classes.

pateash replied:
```suggestion
class BaseDatabricksHook(BaseHook):
```

pateash commented on airflow/providers/databricks/hooks/databricks.py:
```diff
@@ -22,25 +22,14 @@
 operators talk to the ``api/2.0/jobs/runs/submit``
```
Does it precisely talks to **2.0** or **2.x**,
I am asking this because 2.1 is already out.

pateash replied:
```suggestion
operators talk to the ``api/2.x/jobs/runs/submit``
```

pateash commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -0,0 +1,170 @@
+        If not specified, it could be specified in the Databricks connection's extra parameters.
+    """
+
+    conn_name_attr = 'databricks_conn_id'
```
These 3 fields are already present  **databricks_base**,
I don't think we should have to redefine them here.

pateash replied:
```suggestion
```

pateash commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -0,0 +1,170 @@
+    """
+
+    conn_name_attr = 'databricks_conn_id'
+    default_conn_name = 'databricks_default'
```
```suggestion
```

pateash commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -0,0 +1,170 @@
+
+    conn_name_attr = 'databricks_conn_id'
+    default_conn_name = 'databricks_default'
+    conn_type = 'databricks'
```
```suggestion
```

mik-laj commented on setup.py:
```diff
@@ -246,6 +246,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
 ]
 databricks = [
     'requests>=2.26.0, <3',
+    'databricks-sql-connector>=1.0.0, <2.0.0',
```
This library has a license defect. We can't use it yet.https://github.com/apache/airflow/issues/21030#issuecomment-1019363435
Do you consider using any alternative e.g. databricks-dbapi?

alexott replied:
That alternative requires an installation of additional components like odbc driver, that isn’t straightforward. I’ll try to clear licensing issue with dev team

alexott replied:
@mik-laj The source code is downloaded when you install package from pypi.  As it was commented in the original issue, there is no github repo yet, but dev team is working on that.

pateash replied:
@mik-laj @alexott, 
I have 2 questions,

1. Is it okay to use a library not having code publically available? I personally think, it is not.
2. As Databricks API 2.1 is already available as OpenAPI, How difficult  it's gonna be in the future to upgrade if we don't have opensource repository.

alexott replied:
1. Code is installed as part of the package, Apache License doesn't require that the code is hosted on github.  Dev team needs some time to create a repo, setup automation, etc.
2. This library isn't related to the API 2.1 at all.

alexott committed:
More improvements in SQL Operator

This includes:
* identifying SQL Endpoint by name
* allow to output results into a CSV/JSON/JSONL file
* fix tests for DatabricksHook
* address most of the comments

alexott committed:
Started to add tests

alexott committed:
attempt to fix type error

alexott committed:
Revert change that is covered by another PR

lennartkats-db commented on airflow/providers/databricks/hooks/databricks_base.py:
```diff
@@ -0,0 +1,373 @@
+        """
+        method, endpoint = endpoint_info
+
+        # TODO: get rid of explicit 'api/' in the endpoint specification
```
You're thinking to move the `api/` string here?

alexott replied:
yes, because this function is always call API, then it makes no sense to have `api/` prefix for all declared endpoints. Then it's easier to work. But it will be bigger rework, as a separate PR

lennartkats-db commented on airflow/providers/databricks/hooks/databricks_base.py:
```diff
@@ -0,0 +1,373 @@
+        return (
+            isinstance(exception, (requests_exceptions.ConnectionError, requests_exceptions.Timeout))
+            or exception.response is not None
+            and exception.response.status_code >= 500
```
Note that we can also retry on 429 rate limit exceeded. (And I realize this is existing code, so this could be a followup, but perhaps we should make this update while we're refactoring this.)

alexott replied:
yes, I've created a [separate issue for that](https://github.com/apache/airflow/issues/21559).

lennartkats-db commented on docs/apache-airflow-providers-databricks/connections/databricks.rst:
```diff
@@ -34,6 +34,7 @@ There are several ways to connect to Databricks using Airflow.
    i.e. add a token to the Airflow connection. This is the recommended method.
 2. Use Databricks login credentials
    i.e. add the username and password used to login to the Databricks account to the Airflow connection.
+   *Note*: it's not supported for ``DatabricksSqlOperator``.
```
```suggestion
   Note that username/password authentication is discouraged and not supported for ``DatabricksSqlOperator``.
```

lennartkats-db commented on docs/apache-airflow-providers-databricks/connections/databricks.rst:
```diff
@@ -81,6 +82,12 @@ Extra (optional)
+
+    * ``http_path``: optional HTTP path of Databricks SQL endpoint or Databricks cluster. See `documentation <https://docs.databricks.com/dev-tools/python-sql-connector.html#get-started>`_.
+    * ``session_configuration``: optional map containing Spark session configuration parameters.
+    * named internal arguments to the ``Connection`` object from ``databricks-sql-connector`` package.
```
Could you add some examples values for especially the first two here? https://docs.databricks.com/dev-tools/python-sql-connector.html#get-started lists a few

alexott replied:
yep, will add, together with example

lennartkats-db commented on docs/apache-airflow-providers-databricks/operators.rst:
```diff
@@ -132,3 +132,44 @@ Note that there is exactly one named parameter for each top level parameter in t
+`Databricks cluster <https://docs.databricks.com/clusters/index.html>`_.
+
+
+Using the Operator
```
Could you provide an example that uses the operator?

alexott committed:
Emphasize that username/password isn't supported for DatabricksSqlOperator

Co-authored-by: Lennart Kats (databricks) <lennart.kats@databricks.com>

alexott committed:
Rework documentation & add examples for Databricks SQL operator

Split documentation for operators into separate pages & add more content
and examples.

alexott committed:
Added test for Databricks SQL operator

alexott committed:
Added DatabricksCopyIntoOperator to import data into Databricks

alexott committed:
Fix pieces left from rebase, also add a doc for new parameter
